core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeTransformationManager.java

```diff
@@ -104,37 +98,127 @@ public void run() {
-        if (rowIds.isEmpty() || messages.isEmpty() || rowIds.size() != messages.size())
-            return null;
-        StringBuilder jsonPayload = new StringBuilder();
-        jsonPayload.append("{");
-        jsonPayload.append("\"batch\" :");
-        jsonPayload.append("[");
-        int totalBatchSize = Utils.getUTF8Length(jsonPayload) + 2;
-        try {
-            for (int i = 0; i < rowIds.size(); i++) {
-                StringBuilder message = new StringBuilder();
-                message.append("{");
-                message.append("\"orderNo\":").append(rowIds.get(i)).append(",");
-                message.append("\"event\":").append(messages.get(i));
-                message.append("}");
-                message.append(",");
-                totalBatchSize += Utils.getUTF8Length(message);
-                if (totalBatchSize >= Utils.MAX_BATCH_SIZE) {
-                    RudderLogger.logDebug(String.format(Locale.US, "DeviceModeTransformationManager: createDeviceTransformPayload: MAX_BATCH_SIZE reached at index: %d | Total: %d", i, totalBatchSize));
-                    break;
-                }
-                jsonPayload.append(message);
+    private void createMessageMap() {
+        for (int i = 0; i < messageIds.size(); i++) {
+            RudderMessage message = gson.fromJson(messages.get(i), RudderMessage.class);
+            RudderTransformationRequest transformationRequest = new RudderTransformationRequest();
+            transformationRequest.setMessage(message);
+            messageMap.put(messageIds.get(i), transformationRequest);
+        }
+    }
+
+    private boolean handleTransformationResponse(Result result, String requestJson) {
+        if (result.status == NetworkResponses.WRITE_KEY_ERROR) {
+            RudderLogger.logDebug("DeviceModeTransformationManager: TransformationProcessor: Wrong WriteKey. Aborting");
+            return true;
+        } else if (result.status == NetworkResponses.NETWORK_UNAVAILABLE) {
+            RudderLogger.logDebug("DeviceModeTransformationManager: TransformationProcessor: Network unavailable. Aborting");
+            return true;
+        } else if (result.status == NetworkResponses.ERROR) {
+            handleError(requestJson);
+        } else if (result.status == NetworkResponses.RESOURCE_NOT_FOUND) { // dumping back the original messages itself to the factories as transformation feature is not enabled
+            handleResourceNotFound(requestJson);
+        } else {
+            handleSuccess(result);
+        }
+        return false;
+    }
+
+    private void handleError(String requestJson) {
+        int delay = Math.min((1 << retryCount) * 500, MAX_DELAY); // Exponential backoff
+        if (retryCount++ == MAX_RETRIES) {
+            retryCount = 0;
+            deviceModeSleepCount = 0;
+            rudderDeviceModeManager.dumpOriginalEvents(gson.fromJson(requestJson, TransformationRequest.class), true);
```

Comment:
You already have the messageMap. Can't it be used instead of deserializing the json?

Reply:
MessageMap dataType was different so it was not possible earlier. But now I've refactored the code further to address this request. Now we no longer need gson to json conversion.